File: apps/android/app/src/main/java/com/exponential/app/ui/issue/TimelineCollapse.kt
package com.exponential.app.ui.issue

import com.exponential.app.data.db.CommentEntity
import com.exponential.app.data.db.IssueEntity
import com.exponential.app.data.db.IssueEventEntity

// One timeline entry — the synthesized "created the issue" item, a human
// comment, or a synced activity event. Moved out of CommentThread.kt (EXP-240)
// so the pure collapse function below is unit-testable without Compose.
internal sealed interface TimelineItem {
    val createdAt: String
    val id: String

    /** Synthesized first item — never collapses, always pinned before the merge. */
    data class Created(val issue: IssueEntity) : TimelineItem {
        override val createdAt get() = issue.createdAt
        override val id get() = "created-${issue.id}"
    }

    data class Comment(val comment: CommentEntity) : TimelineItem {
        override val createdAt get() = comment.createdAt
        override val id get() = comment.id
    }

    data class Event(val event: IssueEventEntity) : TimelineItem {
        override val createdAt get() = event.createdAt
        override val id get() = event.id
    }
}

// A renderable timeline row after collapsing: either one item, or a run of
// consecutive events folded behind a "Show N activity items" expander.
internal sealed interface TimelineRow {
    data class Single(val item: TimelineItem) : TimelineRow

    /** [runKey] = the run's FIRST event id — stable across sync re-emits. */
    data class CollapsedRun(val runKey: String, val count: Int) : TimelineRow
}

/**
 * Fold runs of MORE THAN TWO consecutive [TimelineItem.Event]s into a single
 * [TimelineRow.CollapsedRun] unless the run's key is in [expandedRuns]. The
 * created item and comments never collapse (they break runs). Pure — the view
 * layer owns [expandedRuns] as remembered state keyed by issue.
 */
internal fun collapseTimeline(
    items: List<TimelineItem>,
    expandedRuns: Set<String>,
): List<TimelineRow> {
    val out = mutableListOf<TimelineRow>()
    var i = 0
    while (i < items.size) {
        val item = items[i]
        if (item !is TimelineItem.Event) {
            out += TimelineRow.Single(item)
            i++
            continue
        }
        var j = i
        while (j < items.size && items[j] is TimelineItem.Event) j++
        val run = items.subList(i, j)
        val runKey = run.first().id
        if (run.size > 2 && runKey !in expandedRuns) {
            out += TimelineRow.CollapsedRun(runKey, run.size)
        } else {
            run.forEach { out += TimelineRow.Single(it) }
        }
        i = j
    }
    return out
}

File: apps/ios/Exponential/UI/Issue/TimelineCollapse.swift
import ExpCore
import Foundation

// Pure timeline shaping for the redesigned Activity section (EXP-240) —
// view-free so the collapse rule stays trivially testable and identical to
// Android's TimelineCollapse.

/// One chronological activity item, pre-merge: the synthesized created row,
/// regular comments, and issue events.
enum TimelineItem: Identifiable {
    /// "«creator» created the issue" — synthesized from the issue row itself
    /// (widget issues have no user creator and read "Feedback widget").
    case created(actorId: String?, createdAt: String, isWidget: Bool)
    case comment(CommentEntity)
    case event(IssueEventEntity)

    var id: String {
        switch self {
        case .created: return "created"
        case .comment(let c): return "c-\(c.id)"
        case .event(let e): return "e-\(e.id)"
        }
    }

    var createdAt: String {
        switch self {
        case .created(_, let at, _): return at
        case .comment(let c): return c.createdAt
        case .event(let e): return e.createdAt
        }
    }
}

/// A display row after collapse: either a plain item or a run of consecutive
/// events folded behind a "Show N activity items" expander.
enum TimelineDisplayRow: Identifiable {
    case item(TimelineItem)
    case collapsedRun(key: String, events: [IssueEventEntity])

    var id: String {
        switch self {
        case .item(let item): return item.id
        case .collapsedRun(let key, _): return "run-\(key)"
        }
    }
}

/// Fold runs of MORE than two consecutive events behind an expander row. The
/// run key is the FIRST event's id, so sync re-emits (rows arriving one by one
/// or re-sorting) never reset an expansion the user already opened. The
/// created item and comments never collapse — they always break a run.
func collapseTimeline(_ items: [TimelineItem], expandedRuns: Set<String>) -> [TimelineDisplayRow] {
    var rows: [TimelineDisplayRow] = []
    var run: [IssueEventEntity] = []

    func flushRun() {
        guard !run.isEmpty else { return }
        let key = run[0].id
        if run.count > 2, !expandedRuns.contains(key) {
            rows.append(.collapsedRun(key: key, events: run))
        } else {
            rows.append(contentsOf: run.map { .item(.event($0)) })
        }
        run = []
    }

    for item in items {
        if case .event(let event) = item {
            run.append(event)
        } else {
            flushRun()
            rows.append(.item(item))
        }
    }
    flushRun()
    return rows
}
